### PR TITLE
KFSPTS-22325 add concur expense v3 processing

### PR DESCRIPTION
**Pull request by jhulslander**
=

### CODE REVIEW

**Review by ajd299 — CHANGES_REQUESTED** [review]
Code looks good, the error handling and logging looks great and should make it easier to maintain. I found some minor typos.

File: src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurExpenseV3ServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        LOG.info("validateConcurExpenseReport, validation step not implemented yet.");
+        ArrayList<String> validationMessages = new ArrayList<>();
+        validationMessages.add("Validation not implemented yet");
+        processingResults.add(
+                new ConcurEventNotificationProcessingResultsDTO(ConcurEventNoticationVersion2EventType.ExpenseReport,
+                        ConcurEventNotificationVersion2ProcessingResults.processingError, fullExpenseReport.getId(),
+                        validationMessages));
+        /*
+         * @todo implement validation here
+         */
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurExpenseV3ServiceImpl.java:71]: yes, it is in the next sprint

File: src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurEventNotificationV2WebserviceServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    protected <T> T convertJsonToConcurDTO(String jsonString, Class<T> dtoType) {
+        ObjectMapper objectMapper = CUJsonUtils.buildObjectMapperUsingDefaultTimeZone();
+        T dto;
+        try {
+            dto = objectMapper.readValue(jsonString, dtoType);
+        } catch (JsonProcessingException e) {
+            LOG.error("convertJsonToConcurDTO, unable to convert json to " + dtoType, e);
+            throw new RuntimeException("convertJsonToConcurDTO, unable to convert roken response to a java dto ", e);
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurEventNotificationV2WebserviceServiceImpl.java:123]: ok

File: src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurEventNotificationV2WebserviceServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        int maxRetryCount = findMaxRetries();
+        int retryCount = 0;
+        while (retryCount < maxRetryCount && jsonResponseString == null) {
+            LOG.info("buildConcurDTOFromEndpoint, trying to build " + dtoType + " from concur endpoint, attempt number "
+                    + retryCount);
+            jsonResponseString = callConcurEndpoint(accessToken, concurEndPoint);
+            retryCount++;
+        }
+        if (StringUtils.isBlank(jsonResponseString)) {
+            throw new RuntimeException("buildConcurClientRequest, Unable to call concur endpoint " + concurEndPoint);
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurEventNotificationV2WebserviceServiceImpl.java:57]: ok

File: src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurEventNotificationV2WebserviceServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    protected Response buildConcurClientRequest(Client client, String accessToken, String concurEndPoint) {
+        URI uri;
+        try {
+            uri = new URI(concurEndPoint);
+        } catch (URISyntaxException e) {
+            LOG.error("buildConcurClientRequest, there was a problem building client request.", e);
+            throw new RuntimeException(
+                    "buildConcurClientRequest, An error occured while building URI: " + concurEndPoint, e);
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurEventNotificationV2WebserviceServiceImpl.java:106]: ok

File: src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurExpenseV3ServiceImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurExpenseV3ServiceImpl.java:71]: but the todo comment should be removed, I shall do so

**Comment by jhulslander** [comment]
I have made the changes requested, this is ready for review again

**Review by ajd299 — APPROVED** [review]
Changes look good, merging

**Pull request merged**